cpp/src/arrow/array/util.cc

```diff
@@ -566,33 +587,67 @@ class NullArrayFactory {
       ARROW_ASSIGN_OR_RAISE(run_ends, MakeArrayFromScalar(*run_end_scalar, 1, pool_));
       ARROW_ASSIGN_OR_RAISE(values, MakeArrayOfNull(type.value_type(), 1, pool_));
     }
-    out_->child_data[0] = run_ends->data();
-    out_->child_data[1] = values->data();
+    out_->child_data = {run_ends->data(), values->data()};
     return Status::OK();
   }
 
   Status Visit(const ExtensionType& type) {
-    out_->child_data.resize(type.storage_type()->num_fields());
-    RETURN_NOT_OK(VisitTypeInline(*type.storage_type(), this));
+    if (presizing_zero_buffer_) {
+      ZeroBufferMustBeAtLeast(
+          GetZeroBufferLength(type.storage_type(), nullable_, length_));
+      return Status::OK();
+    }
+
+    ARROW_ASSIGN_OR_RAISE(auto out,
+                          CreateRelated(type.storage_type(), nullable_, length_));
+    *out_ = std::move(*out);
```

Comment:
This looks weird. Is it guaranteed that the raw pointer `out_` is/will be tracked by a `shared_ptr` at some point? Can't we have `out_` be a `shared_ptr` itself?